qiskit/ignis/experiments/calibration/methods/basic_1q.py

```diff
@@ -151,3 +155,88 @@ def __init__(self,
                          workflow=data_processing)
+
+
+class SinglePulseSingleParameterGenerator(Generator):
+    """
+    A generator that generates single pulses and scans a single parameter
+    of that pulse. Note that the same pulse may be applied to multiple qubits
+    to perform simultaneous calibration.
+    """
+
+    def __init__(self,
+                 qubits: Union[int, List[int]],
+                 parameters: Dict,
+                 amplitudes: List,
+                 scanned_parameter: str,
+                 pulse: Callable = None):
+        """
+        Args:
+            qubits: List of qubits to which this calibration can be applied.
+            parameters: The arguments for the pulse schedule.
+            amplitudes: A list of amplitudes for which to generate a circuit.
+            scanned_parameter: Name of the scanned parameter. If it is not present
+                in the parameters dictionary as a ParameterExpression it will
+                be added to this dict.
+            pulse: A parametric pulse function used to generate the schedule
+                that is added to the circuits. This defaults to Gaussian and
+                parameters should contain 'duration', and 'sigma'.
+        """
+        super().__init__('Single pulse generator', qubits)
+        self._pulse = pulse
+        self.parameters = parameters
+        self.amplitudes = amplitudes
+        self.scanned_parameter = scanned_parameter
+
+        # Add the parameter to be scanned if not supplied.
+        if scanned_parameter not in self.parameters:
+            self.parameters[scanned_parameter] = Parameter('α')
+        elif not isinstance(self.parameters[scanned_parameter], ParameterExpression):
+            self.parameters[scanned_parameter] = Parameter('α')
+
+        # Define the QuantumCircuit that this generator will use.
+        self.qc = QuantumCircuit(max(self._qubits)+1, max(self._qubits)+1)
+        for qubit in self._qubits:
+            gate = Gate(scanned_parameter, 1, [self.parameters[scanned_parameter]])
+            self.qc.append(gate, [qubit])
+            self.qc.add_calibration(gate, [qubit], self._schedule(qubit))
+
+        self.qc.measure(self._qubits, self._qubits)
+
+    def circuits(self) -> List[QuantumCircuit]:
+        """
+        Return a list of experiment circuits.
+        This function is also responsible for adding the
+        meta data to the circuits.
+        """
+        return [self.qc.assign_parameters({self.parameters[self.scanned_parameter]: val})
+                for val in self.amplitudes]
+
+    def _schedule(self, qubit: int) -> Schedule:
+        """
+        Creates the schedules that will be added to the circuit.
+
+        Args:
+            qubit: The qubit to which this schedule is applied.
+        """
+        sched = Schedule(name=self.scanned_parameter)
+        ch = DriveChannel(qubit)
+
+        sched += sched.insert(0, SetFrequency(self.parameters.get('frequency', 0.), ch))
```

Comment:
we can use `.pop` here

Reply:
and we also need to recover phase and frequency shift after the pulse otherwise they affects following pulses.

Reply:
I did not want to use pop here because I did not want to assume that `phase` and `frequency` are always in the dict. E.g. a user doing Rabi should not have to worry about `phase` and `frequency`.